packages/notebook/src/actions.tsx

```diff
@@ -2201,51 +2201,67 @@ namespace Private {
             });
             return Promise.resolve(false);
           }
+          let promise = Promise.resolve();
           if (sessionContext.hasNoKernel) {
-            void sessionContextDialogs.selectKernel(sessionContext);
-            return Promise.resolve(false);
-          }
-          const deletedCells = notebook.model?.deletedCells ?? [];
-          executionScheduled.emit({ notebook, cell });
-          return CodeCell.execute(cell as CodeCell, sessionContext, {
-            deletedCells,
-            recordTiming: notebook.notebookConfig.recordTiming
-          })
-            .then(reply => {
-              deletedCells.splice(0, deletedCells.length);
-              if (cell.isDisposed) {
-                return false;
+            promise = sessionContext.startKernel().then(shouldSelect => {
+              if (shouldSelect) {
+                return sessionContextDialogs.selectKernel(sessionContext);
               }
+            });
+          }
```

Comment:
Now we conditionally change promise to start new default kernel or prompt dialogs. Code will be executed after the kernel selection